src/dnaio/_core.pyx

```diff
@@ -399,21 +400,30 @@ def record_names_match(header1: str, header2: str):
     # find the spaces and tabs easily.
     cdef char * header1chars = <char *>PyUnicode_1BYTE_DATA(header1)
     cdef char * header2chars = <char *>PyUnicode_1BYTE_DATA(header2)
-    return record_name_bytes_match(header1chars, header2chars)
+    cdef size_t header2length = <size_t>PyUnicode_GET_LENGTH(header2)
+    return record_name_bytes_match(header1chars, header2chars, header2length)
 
 
-cdef bint record_name_bytes_match(char *header1chars, char *header2chars):
+cdef bint record_name_bytes_match(char *header1chars, char *header2chars,
+                                  size_t header2length):
     """
-    Check whether the ascii-encoded names match.
+    Check whether the ascii-encoded names match. Only header2length is needed.
     """
     # Only the first part (i.e. the name without the comment) is of interest.
     # Find the first tab or space, if not present, strcspn will return the
     # position of the terminating NULL byte. (I.e. the length).
     cdef size_t header1_ends = strcspn(header1chars, b' \t')
-    cdef size_t header2_ends = strcspn(header2chars, b' \t')
-    # Quick check if the lengths match
-    if header1_ends != header2_ends:
-        return False
+
+    if header1_ends > header2length:
+        return False  # Header2's name is shorter.
+
+    # No need to search whitespace in header2. We can check if there is
+    # whitespace or a 0-byte at the expected position
+    # Use optimized ordering of and statements. Most likely byte is 0. (No
+    # comments in the header). So that will end the comparison immediately.
```

Comment:
All Illumina paired-end reads have a comment, so I think this isn’t accurate. My suggestion would be to just remove these two comment lines – unless you benchmark it and find a difference between checking for `'\0'` or `' '` first.

Reply:
Not in my illumina test data, but that is still Hiseq 4000 data. Anyway, I removed the comment as you suggested. You are probably correct there is little optimization value whether 1, 2 or 3 characters need to be matched.

Reply:
You’re right, it depends on the CASAVA version used. I mainly see data produced by CASAVA 1.8+, which [produces read headers with comments](https://support.illumina.com/help/BaseSpace_OLH_009008/Content/Source/Informatics/BS/FileFormat_FASTQ-files_swBS.htm).